src/auth/downscopedclient.ts

```diff
@@ -0,0 +1,254 @@
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+import {GaxiosOptions, GaxiosPromise, GaxiosResponse} from 'gaxios';
+
+import {BodyResponseCallback} from '../transporters';
+import {Credentials} from './credentials';
+import {AuthClient} from './authclient';
+
+import {GetAccessTokenResponse, Headers, RefreshOptions} from './oauth2client';
+import * as sts from './stscredentials';
+
+/**
+ * The required token exchange grant_type: rfc8693#section-2.1
+ */
+const STS_GRANT_TYPE = 'urn:ietf:params:oauth:grant-type:token-exchange';
+/**
+ * The requested token exchange requested_token_type: rfc8693#section-2.1
+ */
+const STS_REQUEST_TOKEN_TYPE = 'urn:ietf:params:oauth:token-type:access_token';
+/**
+ * The requested token exchange subject_token_type: rfc8693#section-2.1
+ */
+const STS_SUBJECT_TOKEN_TYPE = 'urn:ietf:params:oauth:token-type:access_token';
+/** The STS access token exchange end point. */
+const STS_ACCESS_TOKEN_URL = 'https://sts.googleapis.com/v1beta/token';
+
+/**
+ * Offset to take into account network delays and server clock skews.
+ */
+export const EXPIRATION_TIME_OFFSET = 5 * 60 * 1000;
+
+/**
+ * Internal interface for tracking the access token expiration time.
+ */
+interface CredentialsWithResponse extends Credentials {
+  res?: GaxiosResponse | null;
+}
+
+/**
+ * Defines an upper bound of permissions available for a GCP credential.
+ */
+interface CredentialAccessBoundary {
+  accessBoundary: {
+    accessBoundaryRules: AccessBoundaryRule[];
+  };
+}
+
+/** Defines an upper bound of permissions on a particular resource. */
+interface AccessBoundaryRule {
+  availablePermissions: string[];
+  availableResource: string;
+  availabilityCondition?: AvailabilityCondition;
+}
+
+/**
+ * An optional condition that can be used as part of a
+ * CredentialAccessBoundary to further restrict permissions.
+ */
+interface AvailabilityCondition {
+  expression: string;
+  title?: string;
+  description?: string;
+}
+
+export class DownscopedClient extends AuthClient {
+  /**
+   * OAuth scopes for the GCP access token to use. When not provided,
+   * the default https://www.googleapis.com/auth/cloud-platform is
+   * used.
+   */
+  private cachedDownscopedAccessToken: CredentialsWithResponse | null;
+  private readonly stsCredential: sts.StsCredentials;
+  public readonly authClient: AuthClient;
+  public readonly credentialAccessBoundary: CredentialAccessBoundary;
+  public readonly eagerRefreshThresholdMillis: number;
+  public readonly forceRefreshOnFailure: boolean;
+
+  constructor(
+    private client: AuthClient,
```

Comment:
Why don't you call it `private authClient: AuthClient`
This way you don't need to call `this.authClient = client;` later.